docs/ota-client-guide/modules/ROOT/pages/add-meta-updater-to-vendors-sdk.adoc

```diff
@@ -0,0 +1,63 @@
+= Add meta-updater features to the vendor's SDK
+ifdef::env-github[]
+
+[NOTE]
+====
+We recommend that you link:https://docs.ota.here.com/ota-client/latest/{docname}.html[view this article in our documentation portal]. Not all of our articles render correctly in GitHub.
+====
+endif::[]
+
+
+To avoid conflicts and bug classes, we require the use of `usrmerge`, even on systems that do not use systemd. In the https://www.yoctoproject.org/docs/2.7/mega-manual/mega-manual.html[Yocto] ecosystem, it is a general best practice to write `do_install()` functions that install to the variable `\{bindir}`, but some recipes still hard-code the location of `/bin`.
+
+If a board vendor’s BSP has recipes like this, they will need to be patched so that the files install to the correct locations. In older Yocto versions, the only way to do this was to create a new BSP-specific layer with the patches, and only include that BSP-specific layer when the corresponding BSP was in use. (Otherwise, the recipes would fail to parse.)
+
+Currently, the https://patchwork.openembedded.org/patch/148271/[dynamic-layers] feature allows us to conditionally patch recipes from BSP layers if and only if they are present.
+
+An example of a patch can be seen https://git.yoctoproject.org/cgit/cgit.cgi/meta-virtualization/tree/recipes-containers/cgroup-lite/cgroup-lite_1.15.bb[here], and this can be included in the meta-updater as the file `dynamic-layers/virtualization-layer/{asterisk}/{asterisk}/{asterisk}.bbappend` containing the following:
+
+[source,bash]
+----
+do_install() {
+	install -d ${D}/${bindir}
+	install -m 0755 ${S}/scripts/cgroups-mount ${D}/${bindir}
+	install -m 0755 ${S}/scripts/cgroups-umount ${D}/${bindir}
+	install -d ${D}${sysconfdir}/init.d
+	install -m 0755 ${WORKDIR}/cgroups-init ${D}${sysconfdir}/init.d/cgroups-init
+	install -d ${D}${systemd_unitdir}/system
+	ln -sf /dev/null ${D}${systemd_unitdir}/system/cgroups-init.service
+}
+----
+
+NOTE: You might need to modify a few installments in the `.bbappend` file, however the file needs to include the complete `do_install()` function.

```

Comment:
Huh?

Reply:
another typo. "install statements" and not "installments".

Reply:
Okay, now I think I see what this is about, and I think there's some domain knowledge you're missing here. A `.bbappend` file is like an amendment to a recipe: you can just modify some marginal parts, or add/change variables, or whatever, without having to fork the recipe or write your own version. Like, take a look at [this bbappend file](https://github.com/advancedtelematic/meta-updater/blob/master/recipes-connectivity/connman/connman_%25.bbappend) in meta-updater: it just adds one line of configuration, and leaves the rest of the recipe alone.

So let's say the `do_install()` function in the original recipe looks like this:

```bash
do_install() {
	install -d ${D}/bin
	install -m 0755 ${S}/scripts/cgroups-mount ${D}/bin
	install -m 0755 ${S}/scripts/cgroups-umount ${D}/bin
	install -d ${D}${sysconfdir}/init.d
	install -m 0755 ${WORKDIR}/cgroups-init ${D}${sysconfdir}/init.d/cgroups-init
	install -d ${D}${systemd_unitdir}/system
	ln -sf /dev/null ${D}${systemd_unitdir}/system/cgroups-init.service
}
```

You might think that, since you only need to fix the install statements that use /bin, your bbappend could do something like this:

```bash
do_install() {
	install -d ${D}/${bindir}
	install -m 0755 ${S}/scripts/cgroups-mount ${D}/${bindir}
	install -m 0755 ${S}/scripts/cgroups-umount ${D}/${bindir}
}
```
This won't work, though, because when you have a function in your bbappend it **replaces** the function from the parent file completely.

The reason for this NOTE to be here is to remind developers that if you want to modify anything inside the `do_install()` function, you have to replace the whole function, even if it's just one line in the function that needs to be modified. I'm not sure that's quite getting across.

Reply:
I think I get it now. I replaced this line with the line from the pdf as I think that probably explains it better.